provisioning/device/src/Transports/Mqtt/ProvisioningTransportHandlerMqtt.cs

```diff
@@ -61,279 +59,350 @@ public class ProvisioningTransportHandlerMqtt : ProvisioningTransportHandler
-                    break;
+                MqttClientPublishResult publishResult = await mqttClient.PublishAsync(message, cancellationToken).ConfigureAwait(false);
+
+                if (publishResult.ReasonCode != MqttClientPublishReasonCode.Success)
+                {
+                    throw new ProvisioningTransportException($"Failed to publish the MQTT registration message with reason code '{publishResult.ReasonCode}'", true);
                 }
-                catch (AggregateException ae)
+
+                RegistrationOperationStatus currentStatus =
+                    await GetTaskCompletionSourceResultAsync(
+                        _checkRegistrationOperationStatusSource,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, $"Current provisioning state: {currentStatus.RegistrationState.Status}.");
+
+                if (currentStatus.RegistrationState.Status != ProvisioningRegistrationStatusType.Assigning)
                 {
-                    ae.Handle((ex) =>
-                    {
-                        if (ex is ConnectException) // We will handle DotNetty.Transport.Channels.ConnectException
-                        {
-                            lastException = ex;
+                    return currentStatus.RegistrationState;
+                }
 
-                            if (Logging.IsEnabled)
-                                Logging.Info(this, $"ConnectException trying to connect to {address}: {ex}");
+                // The service is expected to return a value signalling how long to wait before polling again, but
+                // the SDK has a default value for when the service does not send that value. Included in this default value
+                // is some jitter to help stagger the requests if multiple provisioning device clients are checking their provisioning
+                // state at the same time.
+                TimeSpan pollingDelay = currentStatus.RetryAfter ?? RetryJitter.GenerateDelayWithJitterForRetry(s_defaultOperationPollingInterval);
 
-                            return true;
-                        }
+                if (Logging.IsEnabled)
+                    Logging.Info(this, $"Polling for the current state again in {pollingDelay.TotalMilliseconds} milliseconds.");
 
-                        return false; // Let anything else stop the application.
-                    });
+                await Task.Delay(pollingDelay, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private MqttClientOptionsBuilder CreateMqttClientOptions(ProvisioningTransportRegisterRequest provisioningRequest)
+        {
+            var mqttClientOptionsBuilder = new MqttClientOptionsBuilder();
+
+            string hostName = provisioningRequest.GlobalDeviceEndpoint;
+
+            if (TransportProtocol == ProvisioningClientTransportProtocol.WebSocket)
```

Comment:
```suggestion
            if (TransportProtocol != ProvisioningClientTransportProtocol.WebSocket)
            {
                // handle non proxy because it is shorter
            }
            else
```